examples/decomposition/plot_kernel_pca.py

```diff
@@ -3,78 +3,245 @@
-
-X, y = make_circles(n_samples=400, factor=.3, noise=.05)
-
-kpca = KernelPCA(kernel="rbf", fit_inverse_transform=True, gamma=10)
-X_kpca = kpca.fit_transform(X)
-X_back = kpca.inverse_transform(X_kpca)
-pca = PCA()
-X_pca = pca.fit_transform(X)
-
-# Plot results
-
-plt.figure()
-plt.subplot(2, 2, 1, aspect='equal')
-plt.title("Original space")
-reds = y == 0
-blues = y == 1
-
-plt.scatter(X[reds, 0], X[reds, 1], c="red",
-            s=20, edgecolor='k')
-plt.scatter(X[blues, 0], X[blues, 1], c="blue",
-            s=20, edgecolor='k')
-plt.xlabel("$x_1$")
-plt.ylabel("$x_2$")
-
-X1, X2 = np.meshgrid(np.linspace(-1.5, 1.5, 50), np.linspace(-1.5, 1.5, 50))
-X_grid = np.array([np.ravel(X1), np.ravel(X2)]).T
-# projection on the first principal component (in the phi space)
-Z_grid = kpca.transform(X_grid)[:, 0].reshape(X1.shape)
-plt.contour(X1, X2, Z_grid, colors='grey', linewidths=1, origin='lower')
-
-plt.subplot(2, 2, 2, aspect='equal')
-plt.scatter(X_pca[reds, 0], X_pca[reds, 1], c="red",
-            s=20, edgecolor='k')
-plt.scatter(X_pca[blues, 0], X_pca[blues, 1], c="blue",
-            s=20, edgecolor='k')
-plt.title("Projection by PCA")
-plt.xlabel("1st principal component")
-plt.ylabel("2nd component")
-
-plt.subplot(2, 2, 3, aspect='equal')
-plt.scatter(X_kpca[reds, 0], X_kpca[reds, 1], c="red",
-            s=20, edgecolor='k')
-plt.scatter(X_kpca[blues, 0], X_kpca[blues, 1], c="blue",
-            s=20, edgecolor='k')
-plt.title("Projection by KPCA")
-plt.xlabel(r"1st principal component in space induced by $\phi$")
-plt.ylabel("2nd component")
-
-plt.subplot(2, 2, 4, aspect='equal')
-plt.scatter(X_back[reds, 0], X_back[reds, 1], c="red",
-            s=20, edgecolor='k')
-plt.scatter(X_back[blues, 0], X_back[blues, 1], c="blue",
-            s=20, edgecolor='k')
-plt.title("Original space after inverse transform")
-plt.xlabel("$x_1$")
-plt.ylabel("$x_2$")
-
-plt.tight_layout()
-plt.show()
+pca = PCA(n_components=None)
+kernel_pca = KernelPCA(
+    n_components=None, kernel="rbf", gamma=10, fit_inverse_transform=True,
+    alpha=0.1)
+
+X_test_pca = pca.fit(X_train).transform(X_test)
+X_test_kernel_pca = kernel_pca.fit(X_train).transform(X_test)
+
+# %%
+fig, axs = plt.subplots(ncols=2, nrows=2, figsize=(8, 8))
+
+axs[0, 0].scatter(X_train[:, 0], X_train[:, 1], c=y_train)
+axs[0, 0].set_ylabel("Feature #1")
+axs[0, 0].set_xlabel("Feature #0")
+axs[0, 0].set_title("Training data")
+
+axs[0, 1].scatter(X_test[:, 0], X_test[:, 1], c=y_test)
+axs[0, 1].set_xlabel("Feature #0")
+axs[0, 1].set_title("Testing data")
+
+axs[1, 0].scatter(X_test_pca[:, 0], X_test_pca[:, 1], c=y_test)
+axs[1, 0].set_ylabel("Principal component #1")
+axs[1, 0].set_xlabel("Principal component #0")
+axs[1, 0].set_title("Projection of testing data\n using PCA")
+
+axs[1, 1].scatter(X_test_kernel_pca[:, 0], X_test_kernel_pca[:, 1], c=y_test)
+axs[1, 1].set_xlabel("Principal component #0")
+axs[1, 1].set_title("Projection of testing data\n using KernelPCA")
```

Comment:
Since the project is of the test data, I think including the training data is a little confusing. What do you think of having a one row with 3 columns: testing data, pca, kernel_pca?